extensions/elytron-security-oauth2/runtime/src/main/java/io/quarkus/elytron/security/oauth2/runtime/oauth/OAuth2AuthMechanism.java

```diff
@@ -0,0 +1,70 @@
+package io.quarkus.elytron.security.oauth2.runtime.oauth;
+
+import static io.undertow.util.Headers.WWW_AUTHENTICATE;
+import static io.undertow.util.StatusCodes.UNAUTHORIZED;
+
+import io.undertow.UndertowLogger;
+import io.undertow.security.api.AuthenticationMechanism;
+import io.undertow.security.api.SecurityContext;
+import io.undertow.security.idm.Account;
+import io.undertow.security.idm.IdentityManager;
+import io.undertow.server.HttpServerExchange;
+
+/**
+ * An AuthenticationMechanism that validates a caller based on a bearer token
+ */
+public class OAuth2AuthMechanism implements AuthenticationMechanism {
+
+    private IdentityManager identityManager;
+
+    public OAuth2AuthMechanism(IdentityManager identityManager) {
+        this.identityManager = identityManager;
+    }
+
+    /**
+     * Extract the Authorization header and validate the bearer token if it exists. If it does, and is validated, this
+     * builds the org.jboss.security.SecurityContext authenticated Subject that drives the container APIs as well as
+     * the authorization layers.
+     *
+     * @param exchange - the http request exchange object
+     * @param securityContext - the current security context that
+     * @return one of AUTHENTICATED, NOT_AUTHENTICATED or NOT_ATTEMPTED depending on the header and authentication outcome.
+     */
+    @Override
+    public AuthenticationMechanismOutcome authenticate(HttpServerExchange exchange, SecurityContext securityContext) {
+        String authHeader = exchange.getRequestHeaders().getFirst("Authorization");
+        String bearerToken = authHeader != null ? authHeader.substring(7) : null;
+        if (bearerToken != null) {
+            try {
+                Oauth2Credential credential = new Oauth2Credential(bearerToken);
+                if (UndertowLogger.SECURITY_LOGGER.isTraceEnabled()) {
+                    UndertowLogger.SECURITY_LOGGER.tracef("Bearer token: %s", credential);
+                }
+                // Install the OAuth2 principal as the caller
+                Account account = identityManager.verify(credential);
+                if (account != null) {
+                    securityContext.authenticationComplete(account, "BEARER_TOKEN", false);
+                    UndertowLogger.SECURITY_LOGGER.debugf("Authenticated credential(%s) for path(%s) with roles: %s",
+                            credential, exchange.getRequestPath(), account.getRoles());
+                    return AuthenticationMechanismOutcome.AUTHENTICATED;
+                } else {
+                    UndertowLogger.SECURITY_LOGGER.info("Failed to authenticate Oauth2 bearer token");
```

Comment:
Is it an info, an error, or something that can happen in a normal behavior?

From the code, it looks like a normal behavior so we shouldn't log anything.

Reply:
This is the same code as in the JWT extension, I prefere to keep it consistent.

Authentication failures are normal, but it's important to be able to know that a failure occurs so I always love to see those logs. A user can disabled them by raising the level to WARN.

In fact, I found the current extension not loggin enought in case of authentication failures (for example, if it cannot access the introspection endpoint there is no error on the log). Maybe I will revisit it later to raise more error (but again it's the same as the JWT extension)